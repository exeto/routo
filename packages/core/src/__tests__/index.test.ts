import { createMemoryHistory } from 'history';

import { createRouter } from '../index';

describe('createRouter', () => {
  const HOME = 'router/HOME';
  const POSTS = 'router/POSTS';
  const POST = 'router/POST';

  const routes = [
    {
      id: HOME,
      path: '/',
    },
    {
      id: POSTS,
      path: '/posts',
    },
    {
      id: POST,
      path: '/posts/:id',
    },
  ];

  describe('getState', () => {
    const history = createMemoryHistory();
    const router = createRouter(routes, { history });

    it('should return initial state', () => {
      expect(router.getState()).toMatchSnapshot();
    });

    it('should return new state after push', () => {
      history.push('/posts/42');
      expect(router.getState()).toMatchSnapshot();
    });

    it('should return new state after replace', () => {
      history.replace('/posts');
      expect(router.getState()).toMatchSnapshot();
    });

    it('should return new state after back', () => {
      history.goBack();
      expect(router.getState()).toMatchSnapshot();
    });

    it('should return new state after forward', () => {
      history.goForward();
      expect(router.getState()).toMatchSnapshot();
    });

    it('should return new state after push with query params', () => {
      history.push('/posts?hello=world');
      expect(router.getState()).toMatchSnapshot();
    });
  });

  describe('subscribe', () => {
    it('should run listener after history change', () => {
      const history = createMemoryHistory();
      const router = createRouter(routes, { history });
      const listener1 = jest.fn();
      const listener2 = jest.fn();

      router.subscribe(listener1);
      router.subscribe(listener2);
      expect(listener1).not.toBeCalled();
      expect(listener2).not.toBeCalled();

      history.push('/posts');
      expect(listener1).toBeCalledTimes(1);
      expect(listener1).toMatchSnapshot();
      expect(listener2).toBeCalledTimes(1);
      expect(listener2).toMatchSnapshot();
    });

    it('should unsubscribe listener', () => {
      const history = createMemoryHistory();
      const router = createRouter(routes, { history });
      const listener1 = jest.fn();
      const listener2 = jest.fn();
      const listener3 = jest.fn();

      router.subscribe(listener1);

      const unsubscribe = router.subscribe(listener2);

      router.subscribe(listener3);

      unsubscribe();
      history.push('/posts');

      expect(listener1).toBeCalledTimes(1);
      expect(listener2).not.toBeCalled();
      expect(listener3).toBeCalledTimes(1);
    });
  });
});
