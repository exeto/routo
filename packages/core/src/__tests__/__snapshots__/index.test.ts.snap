// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRouter getState should return initial state 1`] = `
Object {
  "action": null,
  "id": "router/HOME",
  "params": Object {},
  "pathname": "/",
  "prev": null,
  "queryParams": Object {},
  "search": "",
}
`;

exports[`createRouter getState should return new state after back 1`] = `
Object {
  "action": "POP",
  "id": "router/HOME",
  "params": Object {},
  "pathname": "/",
  "prev": Object {
    "action": "REPLACE",
    "id": "router/POSTS",
    "params": Object {},
    "pathname": "/posts",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {},
  "search": "",
}
`;

exports[`createRouter getState should return new state after forward 1`] = `
Object {
  "action": "POP",
  "id": "router/POSTS",
  "params": Object {},
  "pathname": "/posts",
  "prev": Object {
    "action": "POP",
    "id": "router/HOME",
    "params": Object {},
    "pathname": "/",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {},
  "search": "",
}
`;

exports[`createRouter getState should return new state after push 1`] = `
Object {
  "action": "PUSH",
  "id": "router/POST",
  "params": Object {
    "id": "42",
  },
  "pathname": "/posts/42",
  "prev": Object {
    "action": null,
    "id": "router/HOME",
    "params": Object {},
    "pathname": "/",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {},
  "search": "",
}
`;

exports[`createRouter getState should return new state after push with query params 1`] = `
Object {
  "action": "PUSH",
  "id": "router/POSTS",
  "params": Object {},
  "pathname": "/posts",
  "prev": Object {
    "action": "POP",
    "id": "router/POSTS",
    "params": Object {},
    "pathname": "/posts",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {
    "hello": "world",
  },
  "search": "?hello=world",
}
`;

exports[`createRouter getState should return new state after replace 1`] = `
Object {
  "action": "REPLACE",
  "id": "router/POSTS",
  "params": Object {},
  "pathname": "/posts",
  "prev": Object {
    "action": "PUSH",
    "id": "router/POST",
    "params": Object {
      "id": "42",
    },
    "pathname": "/posts/42",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {},
  "search": "",
}
`;

exports[`createRouter subscribe should run listener after history change 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "action": "PUSH",
        "id": "router/POSTS",
        "params": Object {},
        "pathname": "/posts",
        "prev": Object {
          "action": null,
          "id": "router/HOME",
          "params": Object {},
          "pathname": "/",
          "prev": null,
          "queryParams": Object {},
          "search": "",
        },
        "queryParams": Object {},
        "search": "",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createRouter subscribe should run listener after history change 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "action": "PUSH",
        "id": "router/POSTS",
        "params": Object {},
        "pathname": "/posts",
        "prev": Object {
          "action": null,
          "id": "router/HOME",
          "params": Object {},
          "pathname": "/",
          "prev": null,
          "queryParams": Object {},
          "search": "",
        },
        "queryParams": Object {},
        "search": "",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
